src/Extractor/SnowflakeQueryFactory.php

```diff
@@ -11,21 +11,43 @@
 
 class SnowflakeQueryFactory extends DefaultQueryFactory
 {
+    public const SEMI_STRUCTURED_TYPES = ['VARIANT' , 'OBJECT', 'ARRAY'];
+
+    private SnowflakeOdbcConnection $connection;
+
+    private SnowflakeMetadataProvider $metadataProvider;
+
     private ?string $incrementalFetchingColType = null;
 
+    public function __construct(
+        SnowflakeOdbcConnection $connection,
+        SnowflakeMetadataProvider $metadataProvider,
+        array $state
+    ) {
+        $this->connection = $connection;
+        $this->metadataProvider = $metadataProvider;
+        parent::__construct($state);
+    }
+
     public function setIncrementalFetchingColType(string $incrementalFetchingColType): self
     {
         $this->incrementalFetchingColType = $incrementalFetchingColType;
         return $this;
     }
 
-    public function createWithCasting(ExportConfig $exportConfig, DbConnection $connection, array $columnInfo): string
+    public function create(ExportConfig $exportConfig, DbConnection $connection): string
     {
-        $sql = array_merge(
-            iterator_to_array($this->createSelectWithCast($connection, $columnInfo)),
-            iterator_to_array($this->createFrom($exportConfig, $connection))
-        );
-        return implode(' ', $sql);
+        $query = parent::create($exportConfig, $this->connection);
+        $columnInfo = $this->metadataProvider->getColumnInfo($query);
+        $objectColumns = array_filter($columnInfo, function ($column): bool {
+            return in_array($column['type'], self::SEMI_STRUCTURED_TYPES);
+        });
+
+        if (empty($objectColumns)) {
+            return $query;
+        }
+
+        return $this->createWithCasting($exportConfig, $this->connection, $columnInfo);
```

Comment:
Celu logiku vytvarania query som presunul do query factory, aby to bolo na jednom mieste.
Je tu bug https://keboola.atlassian.net/browse/COM-749
ktory fixnem v dalsom PR.